## Project - 2(Advanced Model Implementation)

In [1]:
import os
os.chdir(r'C:\\Users\\sharu\\Downloads')
os.getcwd()

'C:\\Users\\sharu\\Downloads'

#### Step - 1(Loading the Dataset)

In [2]:
import pandas as pd
df = pd.read_csv('Housing.csv')

In [3]:
#Displaying the First Five Rows in the dataset
df.head()

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished

In [4]:
#Displaying the Toatal Number of Rows and Columns
print("The Total Number of Rows are:",df.shape[0])
print("The Total Number of Columns are:",df.shape[1])

The Total Number of Rows are: 545
The Total Number of Columns are: 13


In [5]:
#Reviewing all the Columns in the Dataset
print("Displaying the all the columns in the dataset are:")
df.columns

Displaying the all the columns in the dataset are:


Index(['price', 'area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'furnishingstatus'],
      dtype='object')

In [6]:
#Displaying the information of the given dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [7]:
#Displaying the statistics summary table of the given dataset
df.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693578  
std      0.861586  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

In [8]:
#Displaying the datatypes of the each column in the given dataset
df.dtypes

price                int64
area                 int64
bedrooms             int64
bathrooms            int64
stories              int64
mainroad            object
guestroom           object
basement            object
hotwaterheating     object
airconditioning     object
parking              int64
prefarea            object
furnishingstatus    object
dtype: object

#### Step-2(Data Preparation and Cleaning)

In [9]:
#Checking for the null values and doing its sum
print("The Sum of all the missing Values are:")
df.isnull().sum()

The Sum of all the missing Values are:


price               0
area                0
bedrooms            0
bathrooms           0
stories             0
mainroad            0
guestroom           0
basement            0
hotwaterheating     0
airconditioning     0
parking             0
prefarea            0
furnishingstatus    0
dtype: int64

In [11]:
#Checking for the duplicate values and doing its sum
print("The Sum of all the duplicate Values are:")
df.duplicated().sum()

The Sum of all the duplicate Values are:


np.int64(0)

In [23]:
#Feature Engineering(Identifying and Seperating the input features and Target Variable)
X = df.drop('price',axis=1)
y = df['price']
print("The Shape of Features is:",X.shape)
print("The Shape of Target is:",y.shape)

The Shape of Features is: (545, 12)
The Shape of Target is: (545,)


#### Step-4(Advanced Model Building)

In [24]:
#one-hot encode all the categorical Variables
cat_co = ['mainroad','guestroom','basement','hotwaterheating','airconditioning','prefarea','furnishingstatus']
X = pd.get_dummies(X,columns=cat_co,drop_first=True,dtype=int)

In [26]:
#Spliting the data into train and test by using the train_test_split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
print(X_train.shape,X_test.shape,y_train.shape,y_test.shape)

(436, 13) (109, 13) (436,) (109,)


#### Step-5(Buiding XGBoost Regression Model)

In [49]:
from xgboost import XGBRegressor
xgb_m = XGBRegressor(n_estimators=100,learning_rate=0.3,max_depth = 3,subsample = 0.8,colsample_bytree = 0.8,random_state = 45)
xgb_m.fit(X_train,y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=0.8, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=None, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=0.3, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=3,
             max_leaves=None, min_child_weight=None, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=100,
             n_jobs=None, num_parallel_tree=None, ...)

#### Step-6(Training and Prediction)

In [50]:
from sklearn.metrics import r2_score,mean_absolute_error
y_pred_xgb = xgb_m.predict(X_test)
r2_xgb = r2_score(y_test,y_pred)
mae_xgb = mean_absolute_error(y_test,y_pred_xgb)
print("XGBoost r2_score is:",round(r2_xgb,2))
print("XGBoost mean absolute error is:",round(mae_xgb,3))

XGBoost r2_score is: 0.59
XGBoost mean absolute error is: 1060327.625


#### Step-7(Comparing the Linear Regression and Advanced Model(XGBoost Regression))

In [51]:
#Building the Linear Regression Model
from sklearn.linear_model import LinearRegression
lrm = LinearRegression()
lrm.fit(X_train,y_train)

LinearRegression()

In [52]:
y_pred_lr = lrm.predict(X_test)
r2_lr = r2_score(y_test,y_pred_lr)
mae_lr = mean_absolute_error(y_test,y_pred_lr)
print("Linear Regression r2_score is:",round(r2_lr,3))
print("Linear Regression r2_score is:",round(mae_lr,3))

Linear Regression r2_score is: 0.653
Linear Regression r2_score is: 970043.404


In [53]:
#Difference between two models 
r2_diff = r2_xgb - r2_lr
mae_diff = mae_lr - mae_xgb
print("Difference between the r2_scores of two models is:",round(r2_diff,2))
print("Difference between the mean absolute error of two models is:",round(mae_diff,3))

Difference between the r2_scores of two models is: -0.06
Difference between the mean absolute error of two models is: -90284.221


#### Model Performance Comparison
- Difference in R² (XGBoost − Linear): **-0.06**(Positive means XGBoost is better)
- Difference in MAE (Linear − XGBoost): **-90,284.221**(Positive means XGBoost has small error)
- So, **Linear Regression has higher R²score and lower MAE** than XGBoost.
- So Finnaly we can say that **Linear Regression** has better performance than XGBoost Model

#### Which Model Performed Better?
- **Better model:** Linear Regression.  
- It explains about **0.06 more** of the variance in house prices.  
- Its predictions are on average **~90,000 price units closer** to the true values.

#### Why Did XGBoost Perform Worse?
- Dataset is **small and not very complex**, so a simple linear model works well.  
- XGBoost has **more parameters** and can **overfit** noise on small data.  
- The chosen XGBoost hyperparameters may **not be well‑tuned** for this dataset.

#### Trade-offs between two approaches in terms of Complexity,Training Time,Interpretability
- **Complexity:** Linear Regression is simple; XGBoost is complex with many hyperparameters.  
- **Training time:** Linear Regression is very fast; XGBoost is slower and heavier.  
- **Interpretability:** Linear Regression is easy to explain; XGBoost is more of a black box.

### Conclusion
- For this house price dataset, **Linear Regression is the better choice**.  
- It is **simpler, more interpretable, faster**, and gave **better accuracy** than XGBoost.